torchgen/gen.py

```diff
@@ -1399,6 +1399,165 @@ def get_native_function_declarations(
+    skip_dispatcher_op_registration: bool,
+    gen_dispatch_helpers: bool,
+) -> List[str]:
+    definitions: List[str] = []
+    ns_definitions: Dict[str, List[str]] = defaultdict(list)
+    anonymous_definitions: Dict[str, List[str]] = defaultdict(list)
+    registrations: Dict[str, Dict[str, List[str]]] = defaultdict(dict)
+    newline = "\n"
+    ns_gen = dest.RegisterDispatchKey(
+        backend_idx,
+        Target.NAMESPACED_DEFINITION,
+        selector,
+        rocm=rocm,
+        class_method_name=None,
+        skip_dispatcher_op_registration=skip_dispatcher_op_registration,
+    )
+    anonymous_gen = dest.RegisterDispatchKey(
+        backend_idx,
+        Target.ANONYMOUS_DEFINITION,
+        selector,
+        rocm=rocm,
+        class_method_name=None,
+        skip_dispatcher_op_registration=skip_dispatcher_op_registration,
+    )
+    reg_gen = dest.RegisterDispatchKey(
+        backend_idx,
+        Target.REGISTRATION,
+        selector,
+        rocm=rocm,
+        class_method_name=None,
+        skip_dispatcher_op_registration=skip_dispatcher_op_registration,
+    )
+    for f in grouped_native_functions:
+        backend_metadata = backend_idx.get_kernel(f)
+        kernel_namespace = (
+            backend_metadata.cpp_namespace
+            if backend_metadata
+            else DEFAULT_KERNEL_NAMESPACE
+        ).replace("::native", "")
+
+        ns_definitions[kernel_namespace].extend(
+            ns_gen(f),
+        )
+        anonymous_definitions[kernel_namespace].extend(
+            anonymous_gen(f),
+        )
+        namespace = (
+            f.namespace if isinstance(f, NativeFunction) else f.functional.namespace
+        )
+        if namespace not in registrations[kernel_namespace]:
+            registrations[kernel_namespace] = defaultdict(list)
+        registrations[kernel_namespace][namespace].extend(
+            reg_gen(f),
+        )
+
+    for kernel_namespace in ns_definitions:
+        if len(ns_definitions[kernel_namespace]) == 0:
+            continue
+        ns_helper = NamespaceHelper(namespace_str=kernel_namespace)
+        registration_body = ""
+        for namespace in registrations[kernel_namespace]:
+            if not registrations[kernel_namespace][namespace]:
+                continue
+            registration_body += f"""
+TORCH_LIBRARY_IMPL({namespace}, {dispatch_key}, m) {{
+    {newline.join(registrations[kernel_namespace][namespace])}
+}};"""
+        definitions.extend(
+            f"""
+{ns_helper.prologue}
+
+// NB: TORCH_LIBRARY_IMPL must be in an anonymous namespace to avoid
+// ambiguity with conflicting identifiers that may have been defined in
+// at namespace already.
+namespace {{
+
+{newline.join(dest.gen_registration_helpers(backend_idx)) if gen_dispatch_helpers else ''}
+
+{newline.join(anonymous_definitions[kernel_namespace])}
+
+{'' if skip_dispatcher_op_registration else registration_body}
+
+}} // anonymous namespace
+
+namespace {dispatch_key.lower()} {{
```

Comment:
so, do we have to worry about multiple namespaces showing up the same file? (or is it always "aten plus one more namespace").

I think I find this whole function a bit harder to reason about than the previous code - before I could look at `RegisterDispatchKey.cpp and see the 3 distinct pieces of logic (anonymous definitions, namespaced definitions, registrations), as well as what namespace they were in (hardcoded to ATen).

In this PR we still want all of the same stuff, except for another namespace to be there. Can we keep more of the information directly in the C++ template file? E.g:
```
namespace ${op_namespace} {
namespace {
${dispatch_helpers}
${dispatch_anonymous_definitions}
${static_init_dispatch_registrations}
} // anonymous namespace
${deferred_dispatch_registrations}
namespace ${dispatch_namespace} {
${dispatch_namespaced_definitions}
} // namespace ${dispatch_namespace}
} // namespace ${op_namespace}
```
(or if "ATen" + "custom namespace" need to be able to coexist in the same file, then we could just stamp out two of these in the template?)